[Federation] Improve the logging and user feedback in 'kubefed init'

perotinus opened the pull request:
This is a follow-up to #41849, which added some status information. This PR is based off of that one, and includes its changes as well.

See #41725.

```release-note
None
```

perotinus committed:
[Federation] Improve the logging and user feedback in 'kubefed init'.

madhusudancs commented:
@k8s-bot federation gce e2e test this

madhusudancs commented:
@k8s-bot ok to test

csbell commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -358,29 +398,56 @@ func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
+
+	return wait.Poll(2*time.Second, initStepTimeout, func() (bool, error) {
+		fmt.Fprint(cmdOut, ".")
+		if !done {
```
The usage of done as a flag is racy (not the most unsafe) but i believe a channel would be cleaner.

Also, what's the implication of the timeout that this PR introduces? First, it would be good to document how 2 minutes is determined to be a safe default. Also, does this new timing out behavior affect kubefed's ability to clean up existing resources?

perotinus replied:
I've changed this to use a channel instead of a flag.

The longest operation that init does usually seems to take about 1 minute with a GCE cluster. I added a TODO to look into this more.

This timeout should not affect kubefed's ability to clean up resources at the moment, because kubefed does not clean up resources at the moment. The mechanism this timeout uses returns a standard error message, so I'd imagine that any cleanup mechanism would handle this the same as any other failure.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42296)
<!-- Reviewable:end -->

perotinus commented:
@csbell @madhusudancs Is this good to go in? It looks like it's just missing an LGTM.

madhusudancs commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

perotinus commented:
@madhusudancs Looks like the tests passed. Can you LGTM when you get the chance?

irfanurrehman commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 459 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhCv1fL2T_WfWNiiph7:-KhCungNgYvbmUmfohVV:b-vtd1wa) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L459)):*
> The mechanism this timeout uses returns a standard error message, so I'd imagine that any cleanup mechanism would handle this the same as any other f

I find the additional complexity of this function call at different places uncalled for.
What I did request you to do originally was some simple function call like below at the beginning of init.
```
c := make(chan bool)
go func() {
		for {
				select {
				case <-c:
					return
				default:
					fmt.Fprint(cmdOut, ".")
					time.Sleep(2*time.Second)
				}
		}
	}()
```
Which then runs in parallel. Rest of the code remains as is.
Your status messages where ever they are sent to cmdOut will have an additional newline to ensure they are printed on a new line.
```
	fmt.Fprint(cmdOut, "\nstatus msg")
```
Your further need to introduce a time out (which defaults to something) and can be a hidden param as specified here https://github.com/kubernetes/kubernetes/issues/43123, which needs to be passed to waitForPods() and waitSrvHealthy() and ensure they don't wait indefinitely.

Hope I make sense.. :-)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 459 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhCv1fL2T_WfWNiiph7:-KhD-7Vd33x9kzxdH60r:b7efeh0) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L459)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

> The mechanism this timeout uses returns a standard error message, so I'd imagine that any cleanup mechanism would handle this the same as any other f

I find the additional complexity of this function call at different places uncalled for.
What I did request you to do originally was some simple function call like below at the beginning of init.
```
c := make(chan bool)
go func() {
		for {
				select {
				case <-c:
					return
				default:
					fmt.Fprint(cmdOut, ".")
					time.Sleep(2*time.Second)
				}
		}
	}()
```
Which then runs in parallel. Rest of the code remains as is.
Your status messages where ever they are sent to cmdOut will have an additional newline to ensure they are printed on a new line.
```
	fmt.Fprint(cmdOut, "\nstatus msg")
```
Your further need to introduce a time out (which defaults to something) and can be a hidden param as specified here https://github.com/kubernetes/kubernetes/issues/43123, which needs to be passed to waitForPods() and waitSrvHealthy() and ensure they don't wait indefinitely.

Hope I make sense.. :-)

</blockquote></details>

Additionally it will ensure the overall code with cleanup logic is cleaner and readable.
We need to get this change in, so that I can rebase https://github.com/kubernetes/kubernetes/pull/42123 on top.. :-)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 459 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhCv1fL2T_WfWNiiph7:-KhNq_P4IyipoLIa5Vj0:b-4trqaj) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L459)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Additionally it will ensure the overall code with cleanup logic is cleaner and readable.
We need to get this change in, so that I can rebase https://github.com/kubernetes/kubernetes/pull/42123 on top.. :-)
</blockquote></details>

Ah, I see what you're saying. Glad I didn't reply before I did.

I'm not sure yet whether I prefer your approach or what I've done here. I like the fact that what I have here makes it clear that the operations might take a while, and that it allows us to more easily set per-operation timeouts. I think it's easier to read this code as it is now, because the the function call that polls and times out is right next to the code that might take a while, whereas a goroutine at the beginning of the file means that you have to keep that bit of code in mind when you're reading the rest of the file.

@madhusudancs @marun @nikhiljindal any thoughts?

The flag I think I'd rather save for a follow-up PR.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296)*
<!-- Sent from Reviewable.io -->

marun commented:
Nits for sale, getcher nits here...!

---

Reviewed 1 of 3 files at r2.
Review status: 1 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[federation/pkg/kubefed/init/init.go, line 285 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYcb1jT_0TM2aC0uPU:-KhYcb1jT_0TM2aC0uPV:b-pldlea) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L285)):*
> ```Go
> 	}
> 
> 	// 1.
> ```

Why is it useful to retain these comments?  I'm not sure how ordinal comments are anything but a maintenance burden anyway, let alone when there is no associated text. 

---

*[federation/pkg/kubefed/init/init.go, line 286 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYcjXKpum63KGVm90u:-KhYcjXKpum63KGVm90v:bge3d0d) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L286)):*
> ```Go
> 
> 	// 1.
> 	fmt.Fprintf(cmdOut, "Creating a namespace %s for federation system components...", i.commonOptions.FederationSystemNamespace)
> ```

Consider formatting the message only once where logging duplication is necessary.

---

*[federation/pkg/kubefed/init/init.go, line 290 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYdZDgRcGWmShcJHTY:-KhYdZDgRcGWmShcJHTZ:b-xspdit) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L290)):*
> ```Go
> 
> 	err = runAndShowProgress(cmdOut, func() error {
> 		_, err := createNamespace(hostClientset, i.commonOptions.FederationSystemNamespace, i.options.dryRun)
> ```

(No action required) That's a nasty thing to do accidentally.  It looks like ``go vet`` supports an experimental ``-shadow`` flag that should catch this sort of thing:

https://golang.org/cmd/vet/#hdr-Shadowed_variables

It doesn't look look like it can be enabled for ``make verify`` by default, though.  Way too much cruft in the tree.

---

*[federation/pkg/kubefed/init/init.go, line 303 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYfL3bhXrpk5zwZJt7:-KhYfL3bhXrpk5zwZJt8:b-v0ct57) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L303)):*
> ```Go
> 	var svc *api.Service
> 	var ips, hostnames []string
> 	var innerErr error
> ```

I find it confusing that an error from ``createService`` is both returned from runAndShowProgress and set as innerErr.  Consider doing one or the other but not both (that may involve localizing the definition of innerErr to the closure).

---

*[federation/pkg/kubefed/init/init.go, line 459 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhCv1fL2T_WfWNiiph7:-KhYkrnPeyfbsAUqsesO:b998m3s) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L459)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Ah, I see what you're saying. Glad I didn't reply before I did.

I'm not sure yet whether I prefer your approach or what I've done here. I like the fact that what I have here makes it clear that the operations might take a while, and that it allows us to more easily set per-operation timeouts. I think it's easier to read this code as it is now, because the the function call that polls and times out is right next to the code that might take a while, whereas a goroutine at the beginning of the file means that you have to keep that bit of code in mind when you're reading the rest of the file.

@madhusudancs @marun @nikhiljindal any thoughts?

The flag I think I'd rather save for a follow-up PR.
</blockquote></details>

@irfanurrehman's suggestion is simple and it should be obvious that any api call has the potential to take a while.  What's the effective difference between a per-operation timeout and having an equivalent global timeout?

I initially though that the proposed approach would be a good way to log the duration of a given lengthy operation, but realized that the glog output can do the same thing without requiring this helper method.  So the only value of polling per api call that I can see is being able to be explicit that code will be killed if it takes too long.  Given that this is a command that the user can kill at any point with keyboard input (e.g. CTRL-C), is that a distinction worth worrying about?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-KhYpA-c1vfILNxHRYhs:bha0yn2)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@marun @irfanurrehman I've rebased this and modified the status printing mechanism. PTAL.

perotinus commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 285 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYcb1jT_0TM2aC0uPU:-KiWOd9PDCAzyumEcOc3:b-pbdqz3) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L285)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why is it useful to retain these comments?  I'm not sure how ordinal comments are anything but a maintenance burden anyway, let alone when there is no associated text. 
</blockquote></details>

Agreed.

---

*[federation/pkg/kubefed/init/init.go, line 303 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYfL3bhXrpk5zwZJt7:-KiWOmhVHEpe2Ygsj4CN:b1la6tz) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L303)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I find it confusing that an error from ``createService`` is both returned from runAndShowProgress and set as innerErr.  Consider doing one or the other but not both (that may involve localizing the definition of innerErr to the closure).
</blockquote></details>

Agreed, but I've changed the implementation here, so I don't think this is relevant anymore.

---

*[federation/pkg/kubefed/init/init.go, line 459 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhCv1fL2T_WfWNiiph7:-KiWOrZaadWfXZjijJt0:b-p4kx4a) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L459)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

@irfanurrehman's suggestion is simple and it should be obvious that any api call has the potential to take a while.  What's the effective difference between a per-operation timeout and having an equivalent global timeout?

I initially though that the proposed approach would be a good way to log the duration of a given lengthy operation, but realized that the glog output can do the same thing without requiring this helper method.  So the only value of polling per api call that I can see is being able to be explicit that code will be killed if it takes too long.  Given that this is a command that the user can kill at any point with keyboard input (e.g. CTRL-C), is that a distinction worth worrying about?


</blockquote></details>

Good points. I've modified this to have an overall status dot printing loop. Let me know what you think!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 290 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYdZDgRcGWmShcJHTY:-KiWOzrspGFC8-srIvJ9:b-vwa51x) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L290)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) That's a nasty thing to do accidentally.  It looks like ``go vet`` supports an experimental ``-shadow`` flag that should catch this sort of thing:

https://golang.org/cmd/vet/#hdr-Shadowed_variables

It doesn't look look like it can be enabled for ``make verify`` by default, though.  Way too much cruft in the tree.
</blockquote></details>

Hmm. This could be one advantage of having Federation in its own tree: more stringent static analysis.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 286 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYcjXKpum63KGVm90u:-KiWPM2l4s9Pvg0UEjik:b-d1qsjp) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L286)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider formatting the message only once where logging duplication is necessary.
</blockquote></details>

Do you mean only for the ones that have formatting? I'm on the fence here, because I think the overlap between the status statements and the log statements is incidental rather than intrinsic, so I think having them duplicate is reasonable. I'm happy to be convinced otherwise, though. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296)*
<!-- Sent from Reviewable.io -->

perotinus commented:
The unit test failure is due to a data race between printing from within and outside of the status update goroutine. I tried to fix it by unifying the printing in init.go, but there is still a race between that and the testing code that I can't suss out.

I'm not sure what's the best way to fix this. It seems like using the logging library for all printing, perhaps at different verbosities, would fix the issue, since it's thread safe, but that isn't in line with the way the rest of kubefed works. Another alternative is to return to the old pattern, where individual calls are wrapped and logged, though I prefer this implementation.

@marun @irfanurrehman thoughts?

marun commented:
Ugh.  I can't see a way to fix this without either wrapping the calls or calling something like ``startProgress`` and ``endProgress`` to similarly manage a goroutine that prints the periods.  I'm sorry for missing that point when I advocated for this problematic approach.

---

Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[federation/pkg/kubefed/init/init.go, line 286 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhYcjXKpum63KGVm90u:-KiXJxzc0JcaFDUtSuTW:bm93wrv) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L286)):*
<details><summary><i>Previously, perotinus (Jonathan MacMillan) wrote…</i></summary><blockquote>

Do you mean only for the ones that have formatting? I'm on the fence here, because I think the overlap between the status statements and the log statements is incidental rather than intrinsic, so I think having them duplicate is reasonable. I'm happy to be convinced otherwise, though. 
</blockquote></details>

I tend to err on the side of no duplication, but this only happens here so that's fine.

---

*[federation/pkg/kubefed/init/init.go, line 263 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KiXJuBw-H3vrDXAMdzS:-KiXJuBw-H3vrDXAMdzT:b-lupqjs) ([raw file](https://github.com/kubernetes/kubernetes/blob/55ce28bfbc7f7ca553ec4768606786b052aa839c/federation/pkg/kubefed/init/init.go#L263)):*
> ```Go
> // for details.
> func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
> ```

Unrelated formatting change

---

*[federation/pkg/kubefed/init/init.go, line 304 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KiXJMwkfU8NSZrwb8R8:-KiXJMwkfU8NSZrwb8R9:b-mvp9ev) ([raw file](https://github.com/kubernetes/kubernetes/blob/55ce28bfbc7f7ca553ec4768606786b052aa839c/federation/pkg/kubefed/init/init.go#L304)):*
> ```Go
> 	go func() {
> 		err := wait.PollImmediate(2*time.Second, initTimeout, func() (bool, error) {
> 			fmt.Fprint(cmdOut, ".")
> ```

What's to stop a period from appearing at the start of a line, e.g.

```
.Creating federation control plane service
```

Not the end of the world, but it would probably annoy me.

---

*[federation/pkg/kubefed/init/init.go, line 324 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KiXJcFuoGV8QCZA2aqi:-KiXJcFuoGV8QCZA2aqj:b-5ejdlv) ([raw file](https://github.com/kubernetes/kubernetes/blob/55ce28bfbc7f7ca553ec4768606786b052aa839c/federation/pkg/kubefed/init/init.go#L324)):*
> ```Go
> 	glog.V(4).Infof("Created service named %s with IP addresses %v, hostnames %v", svc.Name, ips, hostnames)
> 
> 	// 3a.
> ```

Oops, missed one

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-KiXKEpZxkDq2cPEfn0p:b-qa6mn3)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Hmm. I'm not sure that having ```startProgress``` and ```endProgress``` calls would fix the race between the test and the UUT though.

I think the logging either has to be done in the same goroutine as the test is run in (i.e., go back to wrapping each operation in a function that does logging), or else we need to use a custom logger that is thread safe and that won't have a data race with the test code. Or else, we need to give up on the progress indicator.

I'm happy to introduce a new logger here: it should be a straightforward enough change.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-Kib1dpp_3nliCcaW0tG:bepgda9)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[federation/pkg/kubefed/init/init.go, line 263 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KiXJuBw-H3vrDXAMdzS:-Kib3N8BkmnHErsCo7WI:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/55ce28bfbc7f7ca553ec4768606786b052aa839c/federation/pkg/kubefed/init/init.go#L263)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Unrelated formatting change
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 324 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KiXJcFuoGV8QCZA2aqi:-Kib3I4TFSTH89E_kAcF:b-ucfknh) ([raw file](https://github.com/kubernetes/kubernetes/blob/55ce28bfbc7f7ca553ec4768606786b052aa839c/federation/pkg/kubefed/init/init.go#L324)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Oops, missed one
</blockquote></details>

Thanks!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296)*
<!-- Sent from Reviewable.io -->

marun commented:
Sorry if it wasn't obvious, I meant that startProgress/endProgress would use a synchronization primitive to achieve a similar result to a new logger or the wrapper previously proposed.  Seems kind of lame to have to require thread safety just for testing, though.  Is that the best option?

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-KieoXMjkNnrvKF21ygE:b-uz62vw)*
<!-- Sent from Reviewable.io -->

perotinus commented:
I agree that it's lame, but I'm not sure there's a way around it without removing the test: there's a data race and the tests check for that, so anything that doesn't synchronize with the printing in the test code somehow will, AFAICT, not pass the tests. I prototyped having all the printing in init.go be synchronized with a mutex, but that didn't fix the test code issue. I'm not even quite sure a custom logger will work.

Unfortunately, I couldn't see the stack trace in the testing code–it couldn't be resolved for some reason–so I can't tell 
exactly what the issue is or if there's a better way to fix it.

Perhaps we should remove the fmt logging altogether, and just use different verbosities of glog logging and tell the user to pass -v=X if they want to see status updates? That would fix this issue.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-KifVko50MY1E3sbvR-Z:b-dsvk7q)*
<!-- Sent from Reviewable.io -->

perotinus commented:
@madhusudancs @irfanurrehman @marun At this point, I think the best way to solve this is to use glog for all messages, and print out a notice at the beginning that, for status updates, enable verbose logging. Does anyone else see a better way?

madhusudancs commented:
Overall comment. I don't think we should print dots asynchronously in a goroutine at all. It's a poor user experience to print dots at arbitrary places in a progress text IMO. 

Almost all the operations we perform are against a Kubernetes API server that return immediately. The controllers on that cluster go off and perform operations asynchronously, but we don't wait for them in most cases. There are a few operations where we stop and wait. That's where users need feedback. And that's where we need to print these dots. But all these waiting operations are already wrapped in a polling loop and we just need to print dots inside these loops, preferably one dot per iteration. And that's it.

On the other hand, if kubernetes API server operations don't return immediately, that's an indication of a problem. No dots must be printed there to indicate to users that something is wrong and `kubefed init` is hung. That will help them to make a call about pressing `Ctrl+c`. 

There is one thing I would add though. All the `creating` statements end with a word now. I would add a trailing ellipsis at the end of those statements.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[federation/pkg/kubefed/init/init.go, line 459 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KhCv1fL2T_WfWNiiph7:-KjU8OqGzytUVErGhRH1:b-kk0lhs) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf649d7b8bfb6714fa55b4ecccf01f1fd78200ad/federation/pkg/kubefed/init/init.go#L459)):*
> Your further need to introduce a time out (which defaults to something) and can be a hidden param as specified here #43123, which needs to be passed to waitForPods() and waitSrvHealthy() and ensure they don't wait indefinitely.

@irfanurrehman I agree we must introduce timeouts. But I don't think that should happen in this PR. That deserves its own PR.

> @madhusudancs @marun @nikhiljindal any thoughts?

My comments are above. I don't think we should do any of these, but instead just print dots in polling loops.

> The flag I think I'd rather save for a follow-up PR.

@perotinus yes please! Separate PR.

---

*[federation/pkg/kubefed/init/init.go, line 88 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KjU1AlUPy8uYB5j0x5R:-KjU1AlUPy8uYB5j0x5S:b-xonrnl) ([raw file](https://github.com/kubernetes/kubernetes/blob/604e63c3e2ba1d3f24d30c97c9b50d53867b178f/federation/pkg/kubefed/init/init.go#L88)):*
> ```Go
> 	// TODO: Determine if this is an appropriate timeout. It's currently based
> 	// on observed behavior with a GCE cluster.
> 	initTimeout = 5 * time.Minute
> ```

5 minutes is low. GCE LB allocations can sometimes take longer. This should be somewhere in the 20 minutes ballpark. I would just get rid of this for now by using `PollInfinite` and address this as part of issue #43123.

---

*[federation/pkg/kubefed/init/init.go, line 301 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KjU65IR2ATrFe00KKnD:-KjU65ISQY3_org2nAP_:b-gpjql6) ([raw file](https://github.com/kubernetes/kubernetes/blob/604e63c3e2ba1d3f24d30c97c9b50d53867b178f/federation/pkg/kubefed/init/init.go#L301)):*
> ```Go
> 
> 	// Display a progress indicator, and exit if the overall init timeout is reached.
> 	go func() {
> ```

See my comment above. This goroutine is not required.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-KjU2OU_f7y-dcWHuGVc:b-j51qne)*
<!-- Sent from Reviewable.io -->

perotinus commented:
Heh. Someone recommended in an earlier PR to remove the ellipses, but with this change, it seems to make sense to add them back.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[federation/pkg/kubefed/init/init.go, line 304 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KiXJMwkfU8NSZrwb8R8:-KjjPcLk6I8Lb5fa2bH8:b-vp8ga4) ([raw file](https://github.com/kubernetes/kubernetes/blob/55ce28bfbc7f7ca553ec4768606786b052aa839c/federation/pkg/kubefed/init/init.go#L304)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

What's to stop a period from appearing at the start of a line, e.g.

```
.Creating federation control plane service
```

Not the end of the world, but it would probably annoy me.
</blockquote></details>

This should be fixed now, based on @madhusudancs's suggestions.

---

*[federation/pkg/kubefed/init/init.go, line 88 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KjU1AlUPy8uYB5j0x5R:-KjjT2976zfs3SjYfgH6:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/604e63c3e2ba1d3f24d30c97c9b50d53867b178f/federation/pkg/kubefed/init/init.go#L88)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

5 minutes is low. GCE LB allocations can sometimes take longer. This should be somewhere in the 20 minutes ballpark. I would just get rid of this for now by using `PollInfinite` and address this as part of issue #43123.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 301 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-KjU65IR2ATrFe00KKnD:-KjjPb-wCLB2ipWc-p-U:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/604e63c3e2ba1d3f24d30c97c9b50d53867b178f/federation/pkg/kubefed/init/init.go#L301)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

See my comment above. This goroutine is not required.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-KjjOgC373dT8u8x2K1W:b-uowj27)*
<!-- Sent from Reviewable.io -->

marun commented:
Much cleaner, in any case!

madhusudancs commented:
@perotinus was it me by any chance? ;-) I am curious to know the rationale, I can't remember what it was.

/approve

I will wait for the presubmits to pass before adding LGTM.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42296#-:-KjjTvCfXjzcJkFCjOaz:b-r0ftat)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@k8s-bot ok to test

madhusudancs commented:
/lgtm
